### PR TITLE
Tab group

### PR DESCRIPTION
**Pull request by ggrimaudDTS**

### CODE REVIEW

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx [hunk 1]
```diff
@@ -18,14 +19,25 @@ export function TabGroup({
...
     tabGroupId,
+    tabs: /** @type {NodeList | []} */ [],
   }));
+  const navigateTab = useCallback((/** @type {HTMLButtonElement | null} */ tab) => {
+    if (tab) {
+      tab?.focus();
+      tab?.click();
+    }
+  }, []);
+  const findCurrentTabIndex = useCallback(() => (tabGroupState.tabs.findIndex((/** @type {HTMLButtonElement | null} */ tab) => tab?.id === `tab-${tabGroupState.tabGroupId}-${tabGroupState.selectedTabId}`)), [tabGroupState]);
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:39]: This is scary because if the Tab's id formation code is ever changed, this will break trying to find a tab. Maybe a comment in both places that they are "copy-paste" code would be helpful to reminder a future developer to keep the two locations in sync? Or maybe make a `tabId()` function that both places call?

Another option would be to track the current tab index instead of all the tabs. The focus() and click() in `navigateTab` would need to be moved to a useEffect in Tab:
```
useEffect(
() => { 
// myTabIndex could be returned from `myTabIndexRef.current = registerTab()`
if (tabIndexInContext === myTabIndexRef.current) { 
myRef.current.focus();
myRef.current.click();
}
},
[tabIndexInContext]
);
```

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx [hunk 2]
```diff
@@ -38,12 +50,39 @@ export function TabGroup({
...
+      isVertical: !!isVertical,
     }),
     [onChange, setTabGroupState, tabGroupState, value]
   );
+
+  useEffect(() => {
+    if (tabGroupRef?.current) {
+      setTabGroupState((draftState) => {
+        // @ts-ignore
+        draftState.tabs = [...tabGroupRef?.current?.querySelectorAll('button[role=tab]') || []];
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:78]: Other places use `register` and `unregister` methods in the state for children to register that they exist instead of querying the javascript directly. You can see an example of this in ComboBoxOption where it calls `registerOption` and `unregisterOption`.

There's also the possibility of storing the current tab index in state instead of all the tabs?

If a tab were added dynamically, this would not pick it up since it only fires once after the first render. Also, would there ever be a tab group inside of a tab group 😲. If so, this JS query would over fetch.

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx [hunk 3]
```diff
@@ -38,12 +50,39 @@ export function TabGroup({
           setTabGroupState((draftState) => { draftState.selectedTabId = tabId; });
         }
       },
+      navigateNext: () => {
+        const index = findCurrentTabIndex();
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:54]: modulo math may make this simpler?
`const nextIndex = (index + 1) % tabGroupState.tabs.length;`

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx [hunk 4]
```diff
@@ -38,12 +50,39 @@ export function TabGroup({
...
+        if (index !== (tabGroupState.tabs.length - 1)) {
+          navigateTab(tabGroupState?.tabs?.[index + 1] || null);
+        } else {
+          navigateTab(tabGroupState?.tabs?.[0] || null);
+        }
+      },
+      navigatePrevious: () => {
+        const index = findCurrentTabIndex();
+        if (index !== 0) {
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:63]: modulo math may make this simpler?
`const nextIndex = (index + tabGroupState.tabs.length - 1) % tabGroupState.tabs.length;`

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx [hunk 5]
```diff
@@ -38,12 +50,39 @@ export function TabGroup({
...
+          navigateTab(tabGroupState?.tabs?.[index - 1] || null);
+        } else {
+          navigateTab(tabGroupState?.tabs?.[tabGroupState.tabs.length - 1] || null);
+        }
+      },
+      isVertical: !!isVertical,
     }),
     [onChange, setTabGroupState, tabGroupState, value]
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:71]: do all these need to be in the dependency list since they don't seem to be all used in this function? and shouldn't navigateTab be in this list?

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx [hunk 6]
```diff
@@ -38,12 +50,39 @@ export function TabGroup({
           setTabGroupState((draftState) => { draftState.selectedTabId = tabId; });
         }
       },
+      navigateNext: () => {
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:53]: functions in objects can just be:
```
navigateNext() {
}
```
instead of:
```
navigateNext: () => {
```

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx [hunk 7]
```diff
@@ -38,12 +50,39 @@ export function TabGroup({
...
+        // @ts-ignore
+        draftState.tabs = [...tabGroupRef?.current?.querySelectorAll('button[role=tab]') || []];
+      });
+    }
+  }, []);
+
   return (
     <TabGroupContext.Provider value={contextValue}>
-      <div className={joinClassNames('tab-group', className)} id={`tab-group-${tabGroupState.tabGroupId}`}>
+      <div className={joinClassNames('tab-group', className, isVertical && 'tab-group--vertical')} id={`tab-group-${tabGroupState.tabGroupId}`} ref={tabGroupRef}>
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:85]: line is getting a little long and hard to see all the props? maybe break it out to separate lines for each prop?

```
<div
className={...}
id={...}
ref={...}
>
```

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroupContext.jsx [hunk 8]
```diff
@@ -6,4 +6,7 @@ export const TabGroupContext = /** @type {typeof createContext<TabGroupContextVa
   selectedTabId: '',
   setSelectedTabId: () => { },
   tabGroupId: '',
+  navigateNext: () => { },
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabGroupContext.jsx:9]: not sorted... 🤷

File: @utahdts/utah-design-system/react/components/containers/tabs/TabList.jsx [hunk 9]
```diff
@@ -8,14 +8,18 @@ import { TabGroupContext } from './TabGroupContext';
  * @param {string} [props.className]
  * @returns {import('react').JSX.Element}
  */
-export function TabList({ children, className }) {
-  const { tabGroupId } = useContext(TabGroupContext);
+export function TabList({
+  children,
+  className,
+}) {
+  const { tabGroupId, isVertical } = useContext(TabGroupContext);
```
Comment (AustinHaws-Utah) [inline @utahdts/utah-design-system/react/components/containers/tabs/TabList.jsx:15]: surprised there isn't already a hook for `useTabGroupContext()`?

File: utah-design-system-website/src/@types/jsDocTypes.d.js [hunk 10]
```diff
@@ -314,6 +314,16 @@
  * }
  */
 
+/**
+ * @typedef TabGroupPropsShape {
+ *  @property {string} tabA
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/@types/jsDocTypes.d.js:319]: not sorted... 🤷

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupDocumentation.jsx [hunk 11]
```diff
@@ -17,14 +20,44 @@ export function TabGroupDocumentation() {
       <hr />
 
       <h2 id="section-example">Example</h2>
+      <SandboxExample
+        componentClassName="sandbox-example__component--outline"
+        defaultProps={{
+          tabA: 'Tab A',
+          tabB: 'Tab B',
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupDocumentation.jsx:27]: not sorted... 🤷

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupExampleCodeReact.jsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      isVertical,
+    },
+  },
+}) {
+  return (
+    <>
+      &lt;TabGroup
+      <ExampleCodeReactProp displayProp={` isVertical="${isVertical}">`} indentLevel={0} />
+      <ExampleCodeReactProp displayProp="<TabList>" indentLevel={1} />
+      <ExampleCodeReactProp displayProp={`<Tab id="button-A">${tabA}</Tab>`} indentLevel={2} />
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupExampleCodeReact.jsx:28]: maybe instead of `button-A` it could be `tab-A`? Might have to change the render and a couple of places for this, but it caught me by surprise to see button for a tab.

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupExampleRender.jsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+      isVertical,
+    },
+  },
+  innerRef,
+}) {
+  return (
+    <div style={{ width: '80%' }} ref={innerRef}>
+      <TabGroup defaultValue="button-A" isVertical={isVertical}>
+        <TabList>
+          <Tab id="button-A">{tabA}</Tab>
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupExampleRender.jsx:28]: here is also where the id is `button` instead of `tab`

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+                <TableCell><code className="primary-color">className</code></TableCell>
+                <TableCell><code>string</code></TableCell>
+                <TableCell>null</TableCell>
+                <TableCell>
+                  The CSS Class to use on the tab group wrapper.
+                </TableCell>
+              </TableRow>
+              <TableRow>
+                <TableCell><code className="primary-color">defaultValue</code></TableCell>
+                <TableCell><code>number</code></TableCell>
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx:44]: I wonder if defaultValue should now be changed to be a `tabId`, instead of a sequence number, since every tab is now required to have an id and that would be safer in case the tabs get rearranged in the composition. This includes changing the type from `number` to `string`.

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+              </TableRow>
+              <TableRow>
+                <TableCell><code className="primary-color">onChange</code></TableCell>
+                <TableCell>
+                  <div className="props-code-wrapper">
+                    <code>function</code>
+                  </div>
+                </TableCell>
+                <TableCell>null</TableCell>
+                <TableCell>The function to call when a tab is clicked.</TableCell>
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx:67]: Maybe add that the function is passed the new selected tab's tabId as the value?

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+              </TableRow>
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx:68]: Missing an entry for the `value` prop?

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx [hunk 17]
```diff
@@ -0,0 +1,277 @@
...
+                <TableCell>
+                  The CSS Class to use on the title.
+                </TableCell>
+              </TableRow>
+              <TableRow>
+                <TableCell><code className="primary-color">tagName</code></TableCell>
+                <TableCell><code>string</code></TableCell>
+                <TableCell><code>div</code></TableCell>
+                <TableCell>
+                  What HTML element should the title be.
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx:112]: This is phrased as a question so should probably end with a question mark? Or maybe be changed to something like: 'The HTML element used to wrap around the title.`

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+                <TableCell><code className="primary-color">children</code></TableCell>
+                <TableCell>
+                  <div className="props-code-wrapper">
+                    <code>node</code>
+                  </div>
+                </TableCell>
+                <TableCell><em>required</em></TableCell>
+                <TableCell>
+                  The content of the panels.<br />
+                  Includes <code>Panel</code>.
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx:221]: Panel should probably be TabPanel instead of Panel.

File: utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+                  The CSS Class to use on the panel wrapper.
+                </TableCell>
+              </TableRow>
+              <TableRow>
+                <TableCell><code className="primary-color">tabId</code></TableCell>
+                <TableCell><code>string</code></TableCell>
+                <TableCell><em>required</em></TableCell>
+                <TableCell>
+                  The corresponding tab id.<br />
+                  Must be matching the <code>Tab</code> <code>id</code> prop.
```
Comment (AustinHaws-Utah) [inline utah-design-system-website/src/react/components/websiteContent/library/components/containers/TabGroup/TabGroupPropsDocumentation.jsx:268]: Maybe add a `'s` to <code>Tab</code>? or rephrase it to something like:
```Must match the <code>id</code> prop of its corresponding <code>Tab</code>.```

File: @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx (hunk 3 above)
Comment (ggrimaudDTS) [reply @utahdts/utah-design-system/react/components/containers/tabs/TabGroup.jsx:54]: 🤯
